paddle/fluid/compiler/piano/layout.cc

```diff
@@ -0,0 +1,49 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/compiler/piano/layout.h"
+#include <algorithm>
+#include <iterator>
+#include "paddle/fluid/string/string_helper.h"
+
+namespace paddle {
+namespace piano {
+
+Layout::Layout(const note::LayoutProto& proto) {
+  minor_to_major_.insert(minor_to_major_.end(), proto.minor_to_major().begin(),
+                         proto.minor_to_major().end());
+}
+
+note::LayoutProto Layout::ToProto() const {
+  note::LayoutProto proto;
+  proto.mutable_minor_to_major()->Reserve(minor_to_major().size());
+  for (auto&& dim : minor_to_major_) {
```

Comment:
why use `auto&&` here? `minor_to_major_` is just `std::vector<int64_t>`, I guess `auto`, `const auto&` are good enough?

Reply:
I use it as an universal reference, but i has changed it to const auto& for clear deduction.